File: pkg/gpu-node-mocker/controllers/streaming_probe.go
/*
Copyright 2026 The KAITO Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"context"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
)

// streamingProbeScript is the probe body, delivered to the probe init container
// via a per-shadow-pod ConfigMap. Same pattern KAITO uses for fetch_sas.py.
//
//go:embed streaming_probe.py
var streamingProbeScript string

// streamingMode classifies how (or whether) an original KAITO pod streams its
// model weights from object storage.
type streamingMode string

const (
	streamingModeNone        streamingMode = "none"
	streamingModeSAS         streamingMode = "sas"
	streamingModeDirectAzure streamingMode = "direct-azure"
	streamingModeUnsupported streamingMode = "unsupported-scheme"
)

// azureBlobScheme and the unsupported schemes below are the object-store URI
// prefixes KAITO passes to vLLM's --model on a streaming pod.
const (
	azureBlobScheme = "az://"
	s3Scheme        = "s3://"
	gcsScheme       = "gs://"
)

// azureWorkloadIdentityEnvVars are injected into the ORIGINAL pod by the
// azure-workload-identity mutating webhook. They are stripped when cloning so
// the webhook re-injects them cleanly into the shadow pod, which makes the
// clone correct whether or not the webhook is idempotent.
var azureWorkloadIdentityEnvVars = map[string]struct{}{
	"AZURE_CLIENT_ID":            {},
	"AZURE_TENANT_ID":            {},
	"AZURE_FEDERATED_TOKEN_FILE": {},
	"AZURE_AUTHORITY_HOST":       {},
}

// workloadIdentityPodLabels are the provider labels that mark a pod as carrying
// a streaming identity. Only these are copied onto the shadow pod.
var workloadIdentityPodLabels = []string{AzureWorkloadIdentityUseLabel}

// detectStreamingMode classifies an original KAITO pod from its spec alone.
//
// Annotations are never consulted: KAITO stamps the streaming annotations on the
// Workspace, and GenerateStatefulSetManifest copies only labels onto the pod
// template, so they never reach the Pod.
//
// Deliberately does NOT look at --load-format. KAITO also sets
// --load-format=runai_streamer on its local-weights path, where the weights
// already sit on the node's disk and --model is a plain directory. Keying on
// that flag would classify such a pod as streaming and send the probe after a
// path the shadow pod does not have.
func detectStreamingMode(pod *corev1.Pod) streamingMode {
	for i := range pod.Spec.InitContainers {
		if pod.Spec.InitContainers[i].Name == KaitoSASFetchInitContainerName {
			return streamingModeSAS
		}
	}
	// Same first-match-across-containers rule as extractModelName.
	switch model := extractModelName(pod); {
	case strings.HasPrefix(model, azureBlobScheme):
		return streamingModeDirectAzure
	case strings.HasPrefix(model, s3Scheme), strings.HasPrefix(model, gcsScheme):
		return streamingModeUnsupported
	default:
		return streamingModeNone
	}
}

// streamingIdentity is the material cloned from a streaming original pod so the
// shadow pod can authenticate to object storage exactly as the real pod would.
type streamingIdentity struct {
	mode           streamingMode
	serviceAccount string
	podLabels      map[string]string
	initContainers []corev1.Container
	volumes        []corev1.Volume
	modelURI       string
	providerEnv    []corev1.EnvVar
	// sasEnvFilePath and sasMountPath are read off the cloned fetch-sas
	// container rather than assumed, so the probe reads the SAS env file from
	// wherever fetch-sas actually writes it.
	sasEnvFilePath string
	sasMountPath   string
}

// buildStreamingIdentity clones the streaming identity from the original pod.
//
// The original pod is bound to a fake node, so its init containers never ran and
// the SAS env file they would have written does not exist. What is cloned is the
// container SPEC; the shadow pod's own copy runs for real and mints a fresh SAS
// into the shadow pod's own volume.
func buildStreamingIdentity(original *corev1.Pod, mode streamingMode) streamingIdentity {
	id := streamingIdentity{mode: mode}

	id.podLabels = map[string]string{}
	for _, key := range workloadIdentityPodLabels {
		if v, ok := original.Labels[key]; ok && v != "" {
			id.podLabels[key] = v
		}
	}

	// Only assume the original's ServiceAccount when the pod is marked as a
	// workload identity. That SA is federated to a real cloud identity, so the
	// clone is deliberately narrowed to pods KAITO already flagged as streaming
	// rather than any SA on any pod.
	if len(id.podLabels) > 0 {
		id.serviceAccount = original.Spec.ServiceAccountName
	}

	switch mode {
	case streamingModeSAS:
		for i := range original.Spec.InitContainers {
			c := original.Spec.InitContainers[i]
			if c.Name != KaitoSASFetchInitContainerName {
				continue
			}
			cloned := sanitizeClonedContainer(*c.DeepCopy())
			id.sasEnvFilePath = envValue(cloned.Env, KaitoSASEnvFileEnvVar)
			id.sasMountPath = mountPath(cloned.VolumeMounts, KaitoSASSharedVolumeName)
			id.initContainers = append(id.initContainers, cloned)
			id.volumes = volumesForMounts(original, cloned.VolumeMounts)
			break
		}
	case streamingModeDirectAzure:
		id.modelURI = extractModelName(original)
		for i := range original.Spec.Containers {
			if v := envValue(original.Spec.Containers[i].Env, "AZURE_STORAGE_ACCOUNT_NAME"); v != "" {
				id.providerEnv = append(id.providerEnv, corev1.EnvVar{
					Name: "AZURE_STORAGE_ACCOUNT_NAME", Value: v,
				})
				break
			}
		}
	}

	return id
}

// sanitizeClonedContainer removes the state the azure-workload-identity webhook
// injected into the original pod, so the webhook can re-inject it into the
// shadow pod from a clean base.
func sanitizeClonedContainer(c corev1.Container) corev1.Container {
	env := make([]corev1.EnvVar, 0, len(c.Env))
	for _, e := range c.Env {
		if _, injected := azureWorkloadIdentityEnvVars[e.Name]; !injected {
			env = append(env, e)
		}
	}
	c.Env = env

	mounts := make([]corev1.VolumeMount, 0, len(c.VolumeMounts))
	for _, m := range c.VolumeMounts {
		if m.Name != AzureIdentityTokenVolumeName {
			mounts = append(mounts, m)
		}
	}
	c.VolumeMounts = mounts
	return c
}

// volumesForMounts returns the original pod's volumes that the given mounts
// reference. Selecting by mount rather than copying spec.Volumes wholesale keeps
// the original's GPU-node storage out of a shadow pod that runs on a CPU node.
func volumesForMounts(original *corev1.Pod, mounts []corev1.VolumeMount) []corev1.Volume {
	wanted := make(map[string]struct{}, len(mounts))
	for _, m := range mounts {
		wanted[m.Name] = struct{}{}
	}
	var out []corev1.Volume
	for i := range original.Spec.Volumes {
		if _, ok := wanted[original.Spec.Volumes[i].Name]; ok {
			out = append(out, *original.Spec.Volumes[i].DeepCopy())
		}
	}
	return out
}

func envValue(env []corev1.EnvVar, name string) string {
	for _, e := range env {
		if e.Name == name {
			return e.Value
		}
	}
	return ""
}

func mountPath(mounts []corev1.VolumeMount, name string) string {
	for _, m := range mounts {
		if m.Name == name {
			return m.MountPath
		}
	}
	return ""
}

// buildProbeContainer returns the probe init container. It runs after any cloned
// credential-bootstrap container and before the simulator, so a streaming
// misconfiguration keeps the shadow pod — and therefore the mocked inference pod
// — from ever going Ready.
func buildProbeContainer(id streamingIdentity, cfg Config, scriptCM string) corev1.Container {
	env := []corev1.EnvVar{
		{Name: "PROBE_MODE", Value: string(id.mode)},
		{Name: "PROBE_TIMEOUT_SECONDS", Value: strconv.Itoa(cfg.StreamingProbeTimeoutSec)},
		{Name: "RUNAI_STREAMER_MEMORY_LIMIT", Value: strconv.FormatInt(cfg.StreamingProbeStreamerMemLimit, 10)},
	}
	env = append(env, id.providerEnv...)

	mounts := []corev1.VolumeMount{
		{Name: scriptCM, MountPath: StreamingProbeMountPath, ReadOnly: true},
	}

	if id.mode == streamingModeSAS {
		env = append(env, corev1.EnvVar{Name: KaitoSASEnvFileEnvVar, Value: id.sasEnvFilePath})
		if id.sasMountPath != "" {
			mounts = append(mounts, corev1.VolumeMount{
				Name: KaitoSASSharedVolumeName, MountPath: id.sasMountPath,
			})
		}
	} else if id.modelURI != "" {
		env = append(env, corev1.EnvVar{Name: "PROBE_MODEL_URI", Value: id.modelURI})
	}

	quantity := func(s, fallback string) resource.Quantity {
		q, err := resource.ParseQuantity(s)
		if err != nil {
			return resource.MustParse(fallback)
		}
		return q
	}
	limits := corev1.ResourceList{
		corev1.ResourceCPU:    quantity(cfg.StreamingProbeCPU, DefaultStreamingProbeCPU),
		corev1.ResourceMemory: quantity(cfg.StreamingProbeMemory, DefaultStreamingProbeMemory),
	}
	memReq := resource.MustParse(DefaultStreamingProbeMemoryRequest)
	if memLimit := limits[corev1.ResourceMemory]; memReq.Cmp(memLimit) > 0 {
		memReq = memLimit
	}
	cpuReq := resource.MustParse(DefaultStreamingProbeCPURequest)
	if cpuLimit := limits[corev1.ResourceCPU]; cpuReq.Cmp(cpuLimit) > 0 {
		cpuReq = cpuLimit
	}
	requests := corev1.ResourceList{
		corev1.ResourceCPU:    cpuReq,
		corev1.ResourceMemory: memReq,
	}

	return corev1.Container{
		Name:            StreamingProbeContainerName,
		Image:           cfg.StreamingProbeImage,
		ImagePullPolicy: corev1.PullIfNotPresent,
		Command:         []string{"/bin/sh", "-c", probeShellCommand(id.mode)},
		Env:             env,
		VolumeMounts:    mounts,
		Resources:       corev1.ResourceRequirements{Requests: requests, Limits: limits},
	}
}

// probeShellCommand builds the probe entrypoint.
//
// The SAS env file is sourced under `set -a`, mirroring KAITO's own
// export_sas_token_for_streaming.sh: it is written as shell-quoted KEY='value'
// lines, and auto-export picks up whatever fetch-sas wrote without this having
// to track the variable names. Unlike that wrapper we fail on a missing file
// rather than continuing, because a probe that skips the credentials is not
// validating anything.
//
// torch is installed FIRST, pinned, and from the CPU-only index. runai-model-
// streamer hard-depends on torch, and resolving it from PyPI would pull ~1.3 GB
// of CUDA wheels onto a CPU node. The second install omits --index-url so it
// resolves from PyPI with torch already satisfied.
//
// The SAS mode installs nothing at all: it lists the container over plain HTTPS
// with the standard library, so it needs neither torch nor the streamer.
func probeShellCommand(mode streamingMode) string {
	var b strings.Builder
	b.WriteString("set -e\n")
	if mode == streamingModeSAS {
		b.WriteString(`if [ ! -f "$STREAM_ENV_FILE" ]; then` + "\n")
		b.WriteString(`  echo "ERROR: SAS env file $STREAM_ENV_FILE not found" >&2; exit 1` + "\n")
		b.WriteString("fi\n")
		b.WriteString("set -a\n")
		b.WriteString(`. "$STREAM_ENV_FILE"` + "\n")
		b.WriteString("set +a\n")
	} else {
		fmt.Fprintf(&b, "pip install --no-cache-dir -q --index-url %s torch==%s\n", TorchCPUIndexURL, TorchVersion)
		fmt.Fprintf(&b, "pip install --no-cache-dir -q 'runai-model-streamer[azure]==%s'\n", RunAIStreamerVersion)
	}
	fmt.Fprintf(&b, "exec python3 %s/%s\n", StreamingProbeMountPath, StreamingProbeScriptFileName)
	return b.String()
}

// probeScriptConfigMapName is content-addressed: the controller may create
// ConfigMaps but not update them, so a changed script has to land on a new
// object rather than silently reuse a stale one.
func probeScriptConfigMapName(shadowName string) string {
	name := fmt.Sprintf("%s-probe-%s", shadowName, hash8(streamingProbeScript))
	if len(name) > 253 {
		name = name[:253]
	}
	return name
}

func hash8(s string) string {
	sum := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])[:8]
}

// ensureProbeScriptConfigMap creates the ConfigMap carrying the probe script.
// The name is content-addressed, so an existing object always has the right
// content and can be reused as-is.
func (r *ShadowPodReconciler) ensureProbeScriptConfigMap(ctx context.Context, namespace, name string, ownerRef metav1.OwnerReference) error {
	existing := &corev1.ConfigMap{}
	if err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, existing); err == nil {
		return nil
	} else if !errors.IsNotFound(err) {
		return fmt.Errorf("get probe configmap %s: %w", name, err)
	}

	cm := &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name:            name,
			Namespace:       namespace,
			Labels:          map[string]string{LabelManagedBy: ControllerName},
			OwnerReferences: []metav1.OwnerReference{ownerRef},
		},
		Data: map[string]string{StreamingProbeScriptFileName: streamingProbeScript},
	}
	if err := r.Create(ctx, cm); err != nil && !errors.IsAlreadyExists(err) {
		return fmt.Errorf("create probe configmap %s: %w", name, err)
	}
	return nil
}

File: pkg/gpu-node-mocker/controllers/streaming_probe.py
# Copyright (c) KAITO authors.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Streamer probe for gpu-node-mocker shadow pods.

Validates the KAITO model-streaming path on CPU, before the inference simulator
starts, so that "mock serving is Ready" also means "streaming is configured
correctly".

Two modes:
  sas           fetch-sas has already minted a SAS; list the container over plain
                HTTPS and assert it holds *.safetensors. Nothing is streamed, so
                this mode needs neither torch nor the streamer.
  direct-azure  list via the streamer and read one tensor into CPU memory, which
                additionally proves bulk transfer works.

Listing is the load-bearing check in both modes. KAITO's own bootstrap does NOT
catch an empty container: a static ModelMirror is marked Ready without inspecting
it, and fetch_sas.py exits 0 when it finds no safetensors. This probe is the only
thing standing between an empty container and a GPU pod that fails to load.

Environment:
  PROBE_MODE                   "sas" or "direct-azure"
  PROBE_MODEL_URI              model URI, used when STREAM_MODEL_URI is unset
  STREAM_MODEL_URI             model URI exported by the sourced SAS env file
  AZURE_STORAGE_ACCOUNT_NAME   storage account, exported by fetch-sas
  AZURE_STORAGE_SAS_TOKEN      SAS token, exported by fetch-sas
  PROBE_TIMEOUT_SECONDS        wall-clock budget for the whole probe
  RUNAI_STREAMER_MEMORY_LIMIT  streamer CPU buffer size in bytes; 0 sizes it to
                               the largest tensor in the shard being read
"""

import os
import sys
import threading
import urllib.error
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET

DEFAULT_TIMEOUT_SECONDS = 600


def _scrub(text: str) -> str:
    """Strip query strings from anything we print.

    A SAS token lives in the query string, so a bare URL in an error message
    would leak a live credential into the pod logs.
    """
    parts = urllib.parse.urlsplit(text)
    if not parts.scheme or not parts.query:
        return text
    return urllib.parse.urlunsplit(
        (parts.scheme, parts.netloc, parts.path, "REDACTED", "")
    )


def _fail(message: str) -> int:
    print(f"ERROR: {_scrub(message)}", file=sys.stderr)
    return 1


def _start_watchdog(seconds: int) -> None:
    """Abort the process if the probe outruns its budget.

    A daemon thread calling os._exit rather than signal.alarm: Python only runs
    signal handlers between bytecodes, so SIGALRM is not reliably delivered
    while the main thread sits inside the streamer's C++ extension. os._exit
    also skips interpreter shutdown, which would otherwise block on those same
    threads.
    """

    def _abort() -> None:
        print(
            f"ERROR: probe exceeded {seconds}s budget; aborting",
            file=sys.stderr,
            flush=True,
        )
        os._exit(1)

    timer = threading.Timer(seconds, _abort)
    timer.daemon = True
    timer.start()


def _list_blobs(account: str, container: str, prefix: str, sas: str) -> list:
    """List blob names under a prefix via the blob REST API.

    Plain HTTPS rather than azure-storage-blob, so the SAS mode installs nothing.
    The endpoint suffix default matches the streamer's own.
    """
    suffix = os.environ.get("AZURE_STORAGE_ENDPOINT_SUFFIX", "core.windows.net")
    url = (
        f"https://{account}.blob.{suffix}/{urllib.parse.quote(container)}"
        f"?restype=container&comp=list&prefix={urllib.parse.quote(prefix)}"
        f"&{sas.lstrip('?')}"
    )
    with urllib.request.urlopen(url, timeout=60) as resp:
        body = resp.read()
    return [e.text or "" for e in ET.fromstring(body).iter("Name")]


def _probe_sas(uri: str) -> int:
    """Validate the SAS path: the token minted, and the container holds weights."""
    account = os.environ.get("AZURE_STORAGE_ACCOUNT_NAME")
    token = os.environ.get("AZURE_STORAGE_SAS_TOKEN")
    if not account or not token:
        return _fail(
            "fetch-sas exported no AZURE_STORAGE_ACCOUNT_NAME/AZURE_STORAGE_SAS_TOKEN; "
            "the SAS mint did not complete"
        )

    parts = urllib.parse.urlsplit(uri)
    container, prefix = parts.netloc, parts.path.lstrip("/")
    try:
        names = _list_blobs(account, container, prefix, token)
    except urllib.error.HTTPError as exc:
        # exc carries the SAS-bearing URL, so report only code and reason.
        return _fail(f"listing {_scrub(uri)} failed: HTTP {exc.code} {exc.reason}")
    except Exception as exc:
        return _fail(f"listing {_scrub(uri)} failed: {type(exc).__name__}")

    tensors = [n for n in names if n.endswith(".safetensors")]
    if not tensors:
        return _fail(
            f"no *.safetensors under {_scrub(uri)} ({len(names)} blob(s) listed) — the "
            "weights are missing or the streaming config points at the wrong container"
        )
    print(f"OK: minted SAS and listed {len(tensors)} safetensors file(s)")
    return 0


def _probe_stream(uri: str) -> int:
    """Validate the workload-identity path by reading one tensor."""
    from runai_model_streamer import SafetensorsStreamer, list_safetensors

    files = list_safetensors(uri)
    if not files:
        return _fail(
            f"no *.safetensors found under {_scrub(uri)} — the model weights are "
            "missing or the streaming configuration points at the wrong container"
        )
    print(f"found {len(files)} safetensors file(s)")

    target = sorted(files)[0]
    print(f"streaming {_scrub(target)} into CPU memory")

    with SafetensorsStreamer() as streamer:
        streamer.stream_file(target, device="cpu", is_distributed=False)
        # get_tensors() is annotated Iterator[torch.tensor] upstream but yields
        # (name, tensor) pairs. One tensor is enough to prove the path works.
        for name, tensor in streamer.get_tensors():
            print(f"OK: read tensor {name} with shape {tuple(tensor.shape)}")
            return 0

    return _fail(f"streamed {_scrub(target)} but read no tensors")


def main() -> int:
    timeout = int(os.environ.get("PROBE_TIMEOUT_SECONDS", DEFAULT_TIMEOUT_SECONDS))
    _start_watchdog(timeout)

    uri = os.environ.get("STREAM_MODEL_URI") or os.environ.get("PROBE_MODEL_URI")
    if not uri:
        return _fail(
            "no model URI: neither STREAM_MODEL_URI (from the SAS env file) "
            "nor PROBE_MODEL_URI is set"
        )

    mode = os.environ.get("PROBE_MODE", "unknown")
    print(f"probing model streaming (mode={mode}, uri={_scrub(uri)})")

    if mode == "sas":
        return _probe_sas(uri)
    return _probe_stream(uri)


if __name__ == "__main__":
    sys.exit(main())
